spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Recurrent.scala

```diff
@@ -497,6 +492,70 @@ object Recurrent extends ContainerSerializable {
     new Recurrent[T](batchNormParams)
   }
 
+  /**
+   * set the cells' output and gradInput to recurrent's output and gradInput
+   * to decrease the copy expense.
+   * Copy src tensor to dst tensor along timeDime, default timeDime 2, batchDim 1
+   * @param src
+   * @param dst
+   */
+  private[bigdl] def copy[@specialized(Float, Double) T: ClassTag](
+    src: ArrayBuffer[Tensor[T]], dst: Tensor[T], offset: Int): Unit = {
+    val dstArr = dst.storage().array()
+    val dstOffset = dst.storageOffset() - 1
+    val batchSize = dst.size(0)
+    val timeSize = dst.size(1)
+    val otherSize = dst.nElement() / (batchSize * timeSize)
+
+    var t = 1
+    val length2 = batchSize * otherSize
+    while ((t + offset) <= timeSize) {
+      val srcArr = src(t - 1).storage().array()
+      val srcOffset = src(t - 1).storageOffset() - 1
+      val length1 = (t - 1) * otherSize + dstOffset
+      var l = 0
+      while (l < length2) {
+        System.arraycopy(srcArr, l + srcOffset, dstArr, timeSize * l + length1, otherSize)
+        l += otherSize
+      }
+      t += 1
+    }
+  }
+
+  /**
+   * exchanges the 1 and 2 dimensions of src tensor, and copy memory to dst
+   * @param src
+   * @param dst
+   */
+  private[bigdl] def transposeMemory[@specialized(Float, Double) T: ClassTag](
+    src: Tensor[T], dst: Tensor[T]): Unit = {
+    val srcSize = src.size()
+    val batchSize = srcSize(0)
+    val timeSize = srcSize(1)
+    val otherSize = src.nElement() / (batchSize * timeSize)
+    val srcArr = src.storage().array()
+    val srcOffset = src.storageOffset() - 1
+
+    srcSize(0) = timeSize
+    srcSize(1) = batchSize
+    dst.resize(srcSize)
+    val dstArr = dst.storage().array()
+    val dstOffset = dst.storageOffset() - 1
+
+    val length3 = timeSize * otherSize
+    var t = 1
+    while (t <= batchSize) {
```

Comment:
It is very complex. Maybe change to sth clearer, like below:
```scala

val recordSize = timeSize * stepSize
val batchStepSize = batchSize * stepSize

var srcOffset = src.storageOffset() – 1
var dstOffset = dst.storageOffset() - 1

var b = 1
while(b <= batchSize) {
  var t = 1
  var l = 0
  var m = 0
  while (t <= timeSize) {
    System.arraycopy(srcArr, srcOffset + l, dstArr, dstOffset + m, stepSize)
    l += stepSize
    m += batchStepSize
    t += 1
  }
  srcOffset  += recordSize 
  dstOffset += stepSize
} 
```
Likewise for the copy method above.

Reply:
And actually it's possible to use copy method to implement the transpose method, so that we don't need two pieces of similar code, yes?